counsel.el

```diff
@@ -4517,6 +4518,53 @@ If there is no such buffer, start a new `shell' with NAME."
                             (reusable-frames . visible)))
     (shell name)))
 
+;;** `counsel-firefox-bookmarks`
+(defvar counsel-firefox-bookmarks-html-file
+  (car (file-expand-wildcards "~/.mozilla/firefox/*/bookmarks.html"))
+  "Firefox's auto exported html bookmarks file.")
+
+(defun counsel-firefox-bookmarks-action (x)
+  "Browse candidate X."
+  (browse-url (get-text-property 0 'href x)))
+
+(defun counsel-firefox-bookmarks--candidates ()
+  (let ((candidates))
+    (if (and counsel-firefox-bookmarks-html-file (file-exists-p counsel-firefox-bookmarks-html-file))
+      (with-temp-buffer
+        (insert-file-contents counsel-firefox-bookmarks-html-file)
+        (goto-char (point-min))
+        (while (re-search-forward "*?<A HREF=\"\\([^\"]+\\)\"[^>]*>\\([^<]+\\)</A>" nil t)
+          (let ((a (match-string 0))
+                (href (match-string 1))
+                (text (match-string 2))
+                (tags nil))
+            (if (string-match "TAGS=\"\\([^\"]+\\)\"" a)
+                (setq tags (match-string 1 a)))
+            (push (propertize (format "%s%s" text (if tags (concat "    :" (replace-regexp-in-string "," ":" tags) ":") "")) 'href href) candidates))))
+      (warn "`counsel-firefox-bookmarks-html-file` not exists"))
```

Comment:
* "not exists" should be "does not exist".
* `warn` should not be used by packages/libraries, as it hard-codes the `emacs` custom group.
* It doesn't really make sense to display a warning before passing an empty collection to `ivy-read`; better to `error` or `signal` a `file-error` with a more descriptive error message and abort completion.